openshift/odo
# Watch command should better handle file synchronization/error reporting when push command is invoked before/after the watch

**What type of PR is this?**
> /kind bug

**What does this PR do / why we need it**:
- Fix the `odo watch` logic such that if you edit a devfile after watch has started, the changes to the devfile will now be picked up.
- Fix the `odo watch` logic such that it now updates the `.odo/odo-file-index.json` file, allowing it to play well with `odo push` if push is called immediately after watch
- Fix the `odo watch` logic to regenerate and use the file index to detect changes when `odo watch` is first invoked (in a particular session), with all subsequent updates during the session using the watched file list for changes
- Adds integration tests that test the behaviours that were failing in the linked issues
- Adds directory delete retry, with exponential backoff, to the `DeleteDir` test utility function, to fix Windows-issues seen by myself and others

**Which issue(s) this PR fixes**:
Fixes #3435
Fixes #3585

Windows test fixes:
Fixes #3371
Fixes #3540

**PR acceptance criteria**:




**How to test changes / Special notes to the reviewer**:

To test #3435:
- Create and push a generic odo project
- Run `odo watch`
- Make a change that invalidates the devfile (for example, replace all instances of `"kind: build"` with `"kind: run"` in the file)
- `odo watch` will now report the same error as `odo push` would, against that devfile

To test #3585:
- Create and `odo push` a generic project
- Change some source file A
- `odo watch` that project
- Change some other source file B
- The k8s pod should include the both changes (source file A and B), and the index file (`.odo/odo-file-index.json`) should include both files
(See #3585 for another scenario you can test, if desired)
